src/data_source/sql.rs

```diff
@@ -0,0 +1,987 @@
+/// have data that should live in the same database as the [`SqlDataSource`] data, you can follow
+/// the steps in [custom migrations](#custom-migrations) to accomodate this. When modifying that
+/// data, you can use [`transaction`](Self::transaction) to atomically synchronize updates to the
+/// other modules' data with updates to the [`SqlDataSource`]. If the additional data is completely
+/// independent of HotShot query service data and does not need to be synchronized, you can also
+/// connect to the database directly to make updates.
+///
+/// In the following example, we compose HotShot query service modules with other application-
+/// specific modules, synchronizing updates using [`transaction`](Self::transaction).
+///
+/// ```
+/// # use async_std::{sync::{Arc, RwLock}, task::spawn};
+/// # use futures::StreamExt;
+/// # use hotshot::types::SystemContextHandle;
+/// # use hotshot_query_service::Error;
+/// # use hotshot_query_service::data_source::{
+/// #   sql::Config, SqlDataSource, UpdateDataSource, VersionedDataSource,
+/// # };
+/// # use hotshot_query_service::testing::mocks::{
+/// #   MockNodeImpl as AppNodeImpl, MockTypes as AppTypes,
+/// # };
+/// # use tide_disco::App;
+/// struct AppState {
+///     hotshot_qs: SqlDataSource,
+///     // additional state for other modules
+/// }
+///
+/// async fn init_server(
+///     config: Config,
+///     mut hotshot: SystemContextHandle<AppTypes, AppNodeImpl>,
+/// ) -> Result<App<Arc<RwLock<AppState>>, Error>, Error> {
+///     let mut hotshot_qs = SqlDataSource::connect(config).await.map_err(Error::internal)?;
+///     // Initialize storage for other modules, using `hotshot_qs` to access the database.
+///     let tx = hotshot_qs.transaction().await.map_err(Error::internal)?;
+///     // ...
+///
+///     let state = Arc::new(RwLock::new(AppState {
+///         hotshot_qs,
+///         // additional state for other modules
+///     }));
+///     let mut app = App::with_state(state.clone());
+///     // Register API modules.
+///
+///     spawn(async move {
+///         let mut events = hotshot.get_event_stream(Default::default()).await.0;
+///         while let Some(event) = events.next().await {
+///             let mut state = state.write().await;
+///             UpdateDataSource::<AppTypes, AppNodeImpl>::update(&mut state.hotshot_qs, &event)
+///                 .await
+///                 .unwrap();
+///             // Update other modules' states based on `event`. Use `hotshot_qs` to include
+///             // database updates in the same atomic transaction as `hotshot_qs.update`.
+///             let tx = state.hotshot_qs.transaction().await.unwrap();
+///
+///             // Commit all outstanding changes to the entire state at the same time.
+///             state.hotshot_qs.commit_version().await.unwrap();
+///         }
+///     });
+///
+///     Ok(app)
+/// }
+/// ```
+#[derive(Debug)]
+pub struct SqlDataSource {
+    client: Client,
+    tx_in_progress: bool,
```

Comment:
Why not Option<Transaction<'>> instead? That way we can simply drop the txn in revert_version and we wouldn't need an extra variable to track whether we have a txn in progress.

Reply:
That's what I wanted to do originally, but because of the lifetime in `Transaction<'a>` you can't store a transaction here in the struct itself, as that would be self-referential.